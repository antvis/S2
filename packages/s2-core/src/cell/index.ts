import { BaseCell } from './base-cell';
import { ColCell } from './col-cell';
import { CornerCell } from './corner-cell';
import { DataCell } from './data-cell';
import { MergedCell } from './merged-cell';
import { RowCell } from './row-cell';
import { TableColCell } from './table-col-cell';
import { TableCornerCell } from './table-corner-cell';
import { TableDataCell } from './table-data-cell';
import { TableRowCell } from './table-series-cell';

export {
  TableCornerCell,
  TableColCell,
  TableRowCell,
  RowCell,
  ColCell,
  DataCell,
  MergedCell,
  CornerCell,
  BaseCell,
  TableDataCell,
};
