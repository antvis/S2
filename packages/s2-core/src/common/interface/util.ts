export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
