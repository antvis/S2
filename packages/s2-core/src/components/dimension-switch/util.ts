import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
