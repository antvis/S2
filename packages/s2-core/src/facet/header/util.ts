import { isNil } from 'lodash';

export const getCellPadding = () => {
  const padding = [12, 4, 12, 4];
  const left = isNil(padding[3]) ? 0 : padding[3];
  const right = isNil(padding[1]) ? 0 : padding[1];
  const top = isNil(padding[0]) ? 0 : padding[0];
  const bottom = isNil(padding[2]) ? 0 : padding[2];
  return {
    left,
    right,
    top,
    bottom,
  };
};
