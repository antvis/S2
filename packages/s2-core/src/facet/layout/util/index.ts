export { default as getDimsConditionByNode } from './get-dims-condition-by-node';
export { default as getColHeight } from './get-col-height';
export { default as getColWidth } from './get-col-width';
export { default as getRowWidth } from './get-row-width';
export { default as processColsList } from './process-cols-list';
export { default as getHeaderHierarchyQuery } from './get-header-hierarchy-query';
