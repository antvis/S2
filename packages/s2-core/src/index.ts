export { Event as GEvent } from '@antv/g-canvas';
export { getTheme } from './theme';
export { Node } from './facet/layout/node';
export { Hierarchy } from './facet/layout/hierarchy';
export { buildTableHierarchy } from './facet/layout/build-table-hierarchy';
export * from './common';
export * from './utils';
export * from './cell';
export * from './sheet-type';
export * from './data-set';
export * from './interaction';
export * from './ui/tooltip';
export * from './facet/header';
