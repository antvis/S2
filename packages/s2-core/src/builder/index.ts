import { deepMix } from '@antv/util';
import {
  CellCallback,
  Conditions,
  CornerHeaderCallback,
  DataCellCallback,
  FrameCallback,
  HierarchyCallback,
  KeepOnlyIds,
  LayoutArrangeCallback,
  LayoutCallback,
  LayoutResultCallback,
  NodeField,
  Pagination,
  SpreadsheetOptions,
  CustomHeaderCells,
  Style,
  Totals,
} from '../common/interface';
import { TREE_ROW_DEFAULT_WIDTH } from '../common/constant';
import { DefaultStyleCfg } from './default-style-cfg';

export class SpreadSheetOptionsBuilder {
  public static Builder(): SpreadSheetOptionsBuilder {
    return new SpreadSheetOptionsBuilder();
  }

  private _conditions: Conditions;

  private _width: number;

  private _height: number;

  private _hierarchyCollapse: boolean;

  private _hierarchyType: 'grid' | 'tree';

  private _linkFieldIds: string[];

  private _pagination: Pagination;

  private _containsRowHeader: boolean;

  private _spreadsheetType: boolean;

  private _style: Partial<Style>;

  private _totals: Totals;

  private _customHeaderCells: CustomHeaderCells;

  private _showSeriesNumber: boolean;

  private _hideNodesIds: string[];

  private _keepOnlyNodesIds: KeepOnlyIds;

  private _registerDefaultInteractions: boolean;

  private _scrollReachNodeField: NodeField;

  private _hideRowColFields: string[];

  private _valueInCols = true;

  // 自定义单元格cell
  private _dataCell: DataCellCallback;

  // 自定义cornerCell
  private _cornerCell: CellCallback;

  // 自定义行头cell
  private _rowCell: CellCallback;

  // 自定义列头cell
  private _colCell: CellCallback;

  // 自定义 frame 边框
  private _frame: FrameCallback;

  // 角头可能需要全部自定义，而不是用交叉表概念的node来渲染
  private _cornerHeader: CornerHeaderCallback;

  // 自定义layout
  private _layout: LayoutCallback;

  // 自定义布局结果
  private _layoutResult: LayoutResultCallback;

  // 行列结构的自定义
  private _hierarchy: HierarchyCallback;

  // 行列 维度值的自定义排序
  private _layoutArrange: LayoutArrangeCallback;

  private _debug: boolean;

  public build(): SpreadsheetOptions {
    return {
      debug: this._debug,
      conditions: this._conditions,
      width: this._width,
      height: this._height,
      hierarchyCollapse: this._hierarchyCollapse || false,
      hierarchyType: this._hierarchyType || 'grid',
      linkFieldIds: this._linkFieldIds || [],
      pagination: this._pagination,
      containsRowHeader: this._containsRowHeader || false,
      spreadsheetType: this._spreadsheetType,
      style: deepMix({}, DefaultStyleCfg(), this._style),
      totals: this._totals,
      customHeaderCells: this._customHeaderCells,
      showSeriesNumber: this._showSeriesNumber || false,
      hideNodesIds: this._hideNodesIds || [],
      keepOnlyNodesIds: this._keepOnlyNodesIds || { rowIds: [], colIds: [] },
      registerDefaultInteractions: this._registerDefaultInteractions,
      scrollReachNodeField: this._scrollReachNodeField || {
        rowField: '',
        colField: '',
      },
      hideRowColFields: this._hideRowColFields || [],
      valueInCols: this._valueInCols || false,
      dataCell: this._dataCell,
      cornerCell: this._cornerCell,
      rowCell: this._rowCell,
      colCell: this._colCell,
      frame: this._frame,
      cornerHeader: this._cornerHeader,
      layout: this._layout,
      layoutResult: this._layoutResult,
      hierarchy: this._hierarchy,
      layoutArrange: this._layoutArrange,
    } as SpreadsheetOptions;
  }

  public debug(debug: boolean) {
    this._debug = debug;
    return this;
  }

  public layoutArrange(layoutArrange: LayoutArrangeCallback) {
    this._layoutArrange = layoutArrange;
    return this;
  }

  public hierarchy(hierarchy: HierarchyCallback) {
    this._hierarchy = hierarchy;
    return this;
  }

  public layoutResult(layoutResult: LayoutResultCallback) {
    this._layoutResult = layoutResult;
    return this;
  }

  public layout(layout: LayoutCallback) {
    this._layout = layout;
    return this;
  }

  public cornerHeader(cornerHeader: CornerHeaderCallback) {
    this._cornerHeader = cornerHeader;
    return this;
  }

  public frame(frame: FrameCallback) {
    this._frame = frame;
    return this;
  }

  public colCell(colCell: CellCallback) {
    this._colCell = colCell;
    return this;
  }

  public rowCell(rowCell: CellCallback) {
    this._rowCell = rowCell;
    return this;
  }

  public cornerCell(cornerCell: CellCallback) {
    this._cornerCell = cornerCell;
    return this;
  }

  public dataCell(dataCell: DataCellCallback) {
    this._dataCell = dataCell;
    return this;
  }

  public valueInCols(valueInCols: boolean) {
    this._valueInCols = valueInCols;
    return this;
  }

  public hideRowColFields(hideRowColFields: string[]) {
    this._hideRowColFields = hideRowColFields;
    return this;
  }

  public scrollReachNodeField(scrollReachNodeField: NodeField) {
    this._scrollReachNodeField = scrollReachNodeField;
    return this;
  }

  public registerDefaultInteractions(registerDefaultInteractions: boolean) {
    this._registerDefaultInteractions = registerDefaultInteractions;
    return this;
  }

  public keepOnlyNodesIds(keepOnlyNodesIds: KeepOnlyIds) {
    this._keepOnlyNodesIds = keepOnlyNodesIds;
    return this;
  }

  public hideNodesIds(hideNodesIds: string[]) {
    this._hideNodesIds = hideNodesIds;
    return this;
  }

  public showSeriesNumber(
    showSeriesNumber: boolean,
  ): SpreadSheetOptionsBuilder {
    this._showSeriesNumber = showSeriesNumber;
    return this;
  }

  public totals(totals: Totals): SpreadSheetOptionsBuilder {
    this._totals = totals;
    return this;
  }

  public style(style: Partial<Style>): SpreadSheetOptionsBuilder {
    this._style = style;
    return this;
  }

  public customHeaderCells(
    customHeaderCells: CustomHeaderCells,
  ): SpreadSheetOptionsBuilder {
    this._customHeaderCells = customHeaderCells;
    return this;
  }

  public spreadsheetType(spreadsheetType: boolean): SpreadSheetOptionsBuilder {
    this._spreadsheetType = spreadsheetType;
    return this;
  }

  public containsRowHeader(
    containsRowHeader: boolean,
  ): SpreadSheetOptionsBuilder {
    this._containsRowHeader = containsRowHeader;
    return this;
  }

  public pagination(pagination: Pagination): SpreadSheetOptionsBuilder {
    this._pagination = pagination;
    return this;
  }

  public linkFieldIds(linkFieldIds: string[]): SpreadSheetOptionsBuilder {
    this._linkFieldIds = linkFieldIds;
    return this;
  }

  public hierarchyType(
    hierarchyType: 'grid' | 'tree',
  ): SpreadSheetOptionsBuilder {
    this._hierarchyType = hierarchyType;
    return this;
  }

  public hierarchyCollapse(
    hierarchyCollapse: boolean,
  ): SpreadSheetOptionsBuilder {
    this._hierarchyCollapse = hierarchyCollapse;
    return this;
  }

  public height(height: number): SpreadSheetOptionsBuilder {
    this._height = height;
    return this;
  }

  public width(width: number): SpreadSheetOptionsBuilder {
    this._width = width;
    return this;
  }

  public conditions(conditions: Conditions): SpreadSheetOptionsBuilder {
    this._conditions = conditions;
    return this;
  }

  private _defaultStyle(): Style {
    return {
      treeRowsWidth: TREE_ROW_DEFAULT_WIDTH,
      collapsedRows: {},
      collapsedCols: {},
      cellCfg: {
        width: 96,
        height: 30,
        padding: 0,
      },
      rowCfg: {
        width: 96,
        widthByField: {},
      },
      colCfg: {
        height: 40,
        widthByFieldValue: {},
        heightByField: {},
        colWidthType: 'adaptive',
        totalSample: 10,
        detailSample: 30,
        showDerivedIcon: true,
        maxSampleIndex: 1,
      },
      plotSize: {
        width: 640,
        height: 480,
      },
      device: 'pc',
    } as Style;
  }
}
