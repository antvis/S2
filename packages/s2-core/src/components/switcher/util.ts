import { filter, flatten, isEmpty, isNil, map } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_PREFIX_CLS,
} from './constant';
import { SwitcherItem, SwitchResult, SwitchState } from './interface';
import { getClassNameWithPrefix } from '@/utils/get-classnames';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (state: SwitchState) => {
  return [
    state[FieldType.Rows],
    state[FieldType.Cols],
    state[FieldType.Values],
  ].reduce((sum, value) => sum + (isEmpty(value) ? 0 : 1), 0);
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const showDimensionCrossRows = (nonEmptyCount: number) =>
  nonEmptyCount < MAX_DIMENSION_COUNT;

export const isMeasureType = (fieldType: FieldType) =>
  fieldType === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitchState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层次项的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked: checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitchState): SwitchResult => {
  // rows and cols can't be hidden
  const rows = map(state[FieldType.Rows], 'id');
  const cols = map(state[FieldType.Cols], 'id');

  // flatten all values and derived values
  const values = flatten(
    map(state[FieldType.Values], (item) => {
      const children = map(item.children, 'id');
      return [item.id, ...children];
    }),
  );

  const filterHiddenValues = (item: SwitcherItem) => item.checked === false;

  //  get all hidden values
  const hiddenValues = flatten(
    map(filter(state[FieldType.Values], filterHiddenValues), (item) => {
      const hiddenChildren = map(
        filter(item.children, filterHiddenValues),
        'id',
      );
      return [item.id, ...hiddenChildren];
    }),
  );

  return { rows, cols, values, hiddenValues };
};
