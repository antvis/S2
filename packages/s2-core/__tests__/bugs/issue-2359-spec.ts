/**
 * @description spec for issue #2359
 * https://github.com/antvis/S2/issues/2359
 * 明细表: 自定义列头误用 dataCell 样式
 */
import { pick } from 'lodash';
import { createTableSheet } from 'tests/util/helpers';
import { TableColCell, drawCustomContent } from '../../src';
import type { S2CellType, S2Options } from '@/common/interface';

class TestColCell extends TableColCell {
  drawTextShape() {
    drawCustomContent(this, {
      values: [['A', 'B', 'C']],
    });
  }
}

const s2Options: S2Options = {
  width: 300,
  height: 480,
  seriesNumber: {
    enable: true,
  },
  colCell: (...args) => new TestColCell(...args),
};

describe('Table Sheet Custom Multiple Values Tests', () => {
  test('should use current cell text theme', async () => {
    const s2 = createTableSheet(s2Options);

    s2.setTheme({
      colCell: {
        measureText: {
          fontSize: 12,
        },
        bolderText: {
          fontSize: 14,
        },
        text: {
          fontSize: 20,
          fill: 'red',
        },
      },
      dataCell: {
        text: {
          fontSize: 30,
          fill: 'green',
        },
      },
    });

    await s2.render();

    const mapTheme = (cell: S2CellType) => {
      return cell
        .getTextShapes()
        .map((shape) => pick(shape.attributes, ['fill', 'fontSize']));
    };

    const colCellTexts = s2.facet.getColCells().map(mapTheme);
    const dataCellTexts = s2.facet.getDataCells().map(mapTheme);

    expect(colCellTexts).toMatchSnapshot();
    expect(dataCellTexts).toMatchSnapshot();
  });
});
