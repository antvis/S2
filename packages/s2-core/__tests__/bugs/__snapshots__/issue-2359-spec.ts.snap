// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Sheet Custom Multiple Values Tests should use current cell text theme 1`] = `
Array [
  Array [
    Object {
      "fill": "#000000",
      "fontSize": 12,
    },
  ],
  Array [
    Object {
      "fill": "red",
      "fontSize": 20,
    },
    Object {
      "fill": "red",
      "fontSize": 20,
    },
    Object {
      "fill": "red",
      "fontSize": 20,
    },
  ],
]
`;

exports[`Table Sheet Custom Multiple Values Tests should use current cell text theme 2`] = `
Array [
  Array [
    Object {
      "fill": "#000000",
      "fontSize": 12,
    },
  ],
  Array [
    Object {
      "fill": "#000000",
      "fontSize": 12,
    },
  ],
  Array [
    Object {
      "fill": "#000000",
      "fontSize": 12,
    },
  ],
  Array [
    Object {
      "fill": "green",
      "fontSize": 30,
    },
  ],
  Array [
    Object {
      "fill": "green",
      "fontSize": 30,
    },
  ],
  Array [
    Object {
      "fill": "green",
      "fontSize": 30,
    },
  ],
]
`;
